### ISSUE TITLE
[express] Add hooks for middleware

### ISSUE THREAD

**Issue by BigsonLvrocha** [DataDog/dd-trace-js#1716]
There is only hooks for request spans, but not for middlewares
I'd like to have that for middlewares as well

I can implement this with a pull request also

**Comment by rochdev** [comment]
What is your use case? I see what you're adding an error in there, are you trying to capture an error from one of your middleware?

**Comment by BigsonLvrocha** [comment]
Yes, the current implementation only uses the fields `name`, `message` and `stack` of the Error object for tagging the span, but I want to have more controle on that by using any extra public fields a custom error might have.

I had a case where message was too generic and a better info on the error was in the field `data` but that was not logged in the trave or in the logs

**Comment by rochdev** [comment]
What is the data type of `data`? If we logged additional fields from error objects when available would that work for your use case?

**Comment by BigsonLvrocha** [comment]
It's string, if logged fields that are string, yes that would help :)

**Comment by BigsonLvrocha** [comment]
but for any fields, there is the problem where it could be anything, even circular references. I don't known how that would be solved

**Comment by rochdev** [comment]
That's already handled by the internal tagger which detects circular dependencies. However, for this use case specifically it will probably make more sense to only allow primitives because there are often complex objects like `req` attached to the error object which would be too verbose.

**Comment by BigsonLvrocha** [comment]
Yes, the option to allow only primitive is a good one, when it's an object it would be a string placholder

**Comment by pinko-fowle** [comment]
> What is your use case? I see what you're adding an error in there, are you trying to capture an error from one of your middleware?

@rochdev In our use case, we are trying to isolate some infrequent memory explosions in our long node.js pipelines. We would love to have a middleware hook that would let us add the ability to capture memory before & after each hook.

Hooks around middleware seem like a pretty fine opportunity. We can write a higher-order function to add some work to our existing middlewares, but this would require changing a lot of code, in a number of affected projects. Just having a hook would be a great general capability for adding some finer instrumentation to our code, and would be a good tool for the dd-trace toolbox.

**Comment by BridgeAR** [comment]
I had a look at this again and there were no upvotes or other reactions, so it seems to be limited to a small user base.

@BigsonLvrocha I believe the following would resolve what you need:

```js
const errorData = new WeakMap()

tracer.use('express', {
  hooks: {
    request (span, req) {
      const data = errorData.get(req)
      if (data !== undefined) span.setTag('error.data', data)
    },
  },
})

app.use((error, req, res, next) => {
  errorData.set(req, error.data)
  next(error)
})
```

@pinko-fowle we implemented a new profiling mechanism `DD_PROFILING_ALLOCATION_ENABLED=true` when `DD_PROFILING_ENABLED=true`. That should provide the required insight as you need it :)

---

With that in mind, I am going to close this as resolved. If you do feel this should be reopened, please reach out to our [support](https://www.datadoghq.com/support/). They can prioritize these things and also provide general support, which we can not provide through github.
